src/identity_tree.rs

```diff
@@ -156,48 +162,48 @@ where
 
     /// Returns _up to_ `maximum_update_count` updates that are to be applied to
     /// the tree.
-    fn peek_next_updates(&self, maximum_update_count: usize) -> Vec<TreeUpdate> {
-        self.next.as_ref().map_or_else(Vec::new, |next| {
-            let next = next.get_data();
-            next.metadata
-                .diff
-                .iter()
-                .take(maximum_update_count)
-                .cloned()
-                .collect()
-        })
+    fn peek_next_updates(&self, maximum_update_count: usize) -> Vec<TreeUpdateWithTree> {
+        let Some(next) = self.next.as_ref() else { return Vec::new(); };
+
+        let next = next.get_data();
+        next.metadata
+            .diff
+            .iter()
+            .take(maximum_update_count)
+            .cloned()
+            .collect()
     }
 
     /// Applies the next _up to_ `update_count` updates, returning the merkle
     /// tree proofs obtained after each apply.
-    fn apply_next_updates(&mut self, update_count: usize) -> Vec<Proof> {
-        let mut proofs: Vec<Proof> = vec![];
-        if let Some(next) = self.next.clone() {
-            {
-                // Acquire the exclusive write lock on the next version.
-                let mut next = next.get_data();
-
-                // Get the updates to be applied and apply them sequentially. It is very
-                // important that we record the merkle proof after each step as they depend on
-                // each other.
-                let updates: Vec<&TreeUpdate> =
-                    next.metadata.diff.iter().take(update_count).collect();
-                for update in &updates {
-                    self.update(update.leaf_index, update.element);
-                    let proof = self.tree.proof(update.leaf_index);
-                    proofs.push(proof);
-                }
-
-                // Remove only the updates that have been consumed, which may be all of them.
-                next.metadata.diff = if next.metadata.diff.len() > updates.len() {
-                    Vec::from(&next.metadata.diff[updates.len()..])
-                } else {
-                    vec![]
-                }
+    fn apply_next_updates(&mut self, update_count: usize) {
+        let Some(next) = self.next.clone() else { return; };
+
+        {
+            // Acquire the exclusive write lock on the next version.
+            let mut next = next.get_data();
+
+            let updates: Vec<&TreeUpdate> = next
+                .metadata
+                .diff
+                .iter()
+                .map(|u| &u.update)
+                .take(update_count)
+                .collect();
+
+            for update in &updates {
```

Comment:
Probably could be smarter about this – no need to recompute diffs and result, could just reuse from the applied updates. We could add something like `apply_diffs(&mut self, diff: Vec<TreeUpdateWithTree>)` to `BasicTreeOps`

Reply:
done